internal/charmstore/addentity.go

```diff
@@ -0,0 +1,648 @@
+}
+
+// denormalizeEntity sets all denormalized fields in e
+// from their associated canonical fields.
+//
+// It is the responsibility of the caller to set e.SupportedSeries
+// if the entity URL does not contain a series. If the entity
+// URL *does* contain a series, e.SupportedSeries will
+// be overwritten.
+func denormalizeEntity(e *mongodoc.Entity) {
+	e.BaseURL = mongodoc.BaseURL(e.URL)
+	e.Name = e.URL.Name
+	e.User = e.URL.User
+	e.Revision = e.URL.Revision
+	e.Series = e.URL.Series
+	if e.URL.Series != "" {
+		if e.URL.Series == "bundle" {
+			e.SupportedSeries = nil
+		} else {
+			e.SupportedSeries = []string{e.URL.Series}
+		}
+	}
+	if e.PromulgatedURL == nil {
+		e.PromulgatedRevision = -1
+	} else {
+		e.PromulgatedRevision = e.PromulgatedURL.Revision
+	}
+}
+
+func (s *Store) newBundle(id *router.ResolvedURL, r io.ReadSeeker, blobSize int64) (charm.Bundle, error) {
+	readerAt := ReaderAtSeeker(r)
+	b, err := charm.ReadBundleArchiveFromReader(readerAt, blobSize)
+	if err != nil {
+		return nil, zipReadError(err, "cannot read bundle archive")
+	}
+	bundleData := b.Data()
+	charms, err := s.bundleCharms(bundleData.RequiredCharms())
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot retrieve bundle charms")
+	}
+	if err := bundleData.VerifyWithCharms(verifyConstraints, verifyStorage, charms); err != nil {
+		// TODO frankban: use multiError (defined in internal/router).
+		return nil, errgo.NoteMask(verificationError(err), "bundle verification failed", errgo.Is(params.ErrInvalidEntity))
+	}
+	return b, nil
+}
+
+func (s *Store) bundleCharms(ids []string) (map[string]charm.Charm, error) {
+	numIds := len(ids)
+	urls := make([]*charm.URL, 0, numIds)
+	idKeys := make([]string, 0, numIds)
+	// TODO resolve ids concurrently.
+	for _, id := range ids {
+		url, err := charm.ParseURL(id)
+		if err != nil {
+			// Ignore this error. This will be caught in the bundle
+			// verification process (see bundleData.VerifyWithCharms) and will
+			// be returned to the user along with other bundle errors.
+			continue
+		}
+		e, err := s.FindBestEntity(url, map[string]int{})
+		if err != nil {
+			if errgo.Cause(err) == params.ErrNotFound {
+				// Ignore this error too, for the same reasons
+				// described above.
+				continue
+			}
+			return nil, err
+		}
+		urls = append(urls, e.URL)
+		idKeys = append(idKeys, id)
+	}
+	var entities []mongodoc.Entity
+	if err := s.DB.Entities().
+		Find(bson.D{{"_id", bson.D{{"$in", urls}}}}).
+		All(&entities); err != nil {
+		return nil, err
+	}
+
+	entityCharms := make(map[charm.URL]charm.Charm, len(entities))
+	for i, entity := range entities {
+		entityCharms[*entity.URL] = &entityCharm{entities[i]}
+	}
+	charms := make(map[string]charm.Charm, len(urls))
+	for i, url := range urls {
+		if ch, ok := entityCharms[*url]; ok {
+			charms[idKeys[i]] = ch
+		}
+	}
+	return charms, nil
+}
+
+// bundleCharms returns all the charm URLs used by a bundle,
+// without duplicates.
+// XXX is this still used?
```

Comment:
It is actually used. It might be possible to integrate Store.bundleCharms and bundleCharms to an extent (and the docs should make the difference clear, but I'd prefer to leave that for another cleanup as this one is big enough already!)